File: Ruby/09_views.rb
#!/usr/bin/env ruby

require "rubygems"
require "couchbase"
require "rainbow"
require "pp"

system 'clear'
puts "--------------------------------------------------------------------------".bright.foreground(:red)
puts "Couchbase Views - List Design Documents and Views".bright.foreground(:red)
puts "--------------------------------------------------------------------------".bright.foreground(:red)
puts

# establish connection to beers bucket
cb = Couchbase.connect(:bucket => "beer-sample")

ddocs = cb.design_docs

# list out each design doc and views in the bucket
ddocs.each do |dd|
  puts "DDoc: " + dd[0].to_s.bright
  this_dd = cb.design_docs[dd[0]]  
  puts "\t" + this_dd.views.inspect  
end
puts
puts

puts "--------------------------------------------------------------------------".bright.foreground(:red)
# Select a View and Query It, output the Document Keys 
puts "Breweries (by_name)".bright.foreground(:red)
puts
breweries = cb.design_docs["breweries"]
breweries.by_name(:limit => 10).each do |doc|
  puts doc.key
end
puts
puts

puts "--------------------------------------------------------------------------".bright.foreground(:red)
# Select a View and Query It, output the Document Keys 
puts "Breweries (by_name), include docs and output country".bright.foreground(:red)
puts
breweries = cb.design_docs["breweries"]
breweries.by_name(:limit => 10, :include_docs => true).each do |doc|
  puts doc.key
  puts "\t" + doc["country"]
end
puts
puts

puts "--------------------------------------------------------------------------".bright.foreground(:red)
# Select a View and Query It, output the Document Keys 
puts "Breweries (by_name) starting with letter y or Y".bright.foreground(:red)
puts
breweries = cb.design_docs["breweries"]
breweries.by_name(:startkey => "y", :endkey => "z").each do |doc|
  puts doc.key
end
puts

puts "--------------------------------------------------------------------------".bright.foreground(:red)
# Select a View and Query It, output the Document Keys 
puts "Breweries (by_name) starting with letter y only, no Y (results should be empty)".bright.foreground(:red)
puts
breweries = cb.design_docs["breweries"]
breweries.by_name(:startkey => "y", :endkey => "Y").each do |doc|
  puts doc.key
end
puts

puts
puts "--------------------------------------------------------------------------".bright.foreground(:red)
puts

File: Python/09_views.py
from couchbase import Couchbase
from couchbase.views.params import Query

cb = Couchbase.connect(bucket='beer-sample')

# Note, this requires the 'by_name' view:
#
# function (doc, meta) {
#   emit(doc.name, null);
# }

print "Querying a view"
rows = cb.query('beer', 'by_name', limit=10)
for row in rows:
    print row

print "Querying views, with documents"
rows = cb.query('beer', 'by_name', include_docs=True, limit=5)
for row in rows:
    print row


print "Querying view with range"
rows = cb.query('beer', 'by_name',
                mapkey_range=['y', 'z'], include_docs=True, limit=5)
for row in rows:
    print row
